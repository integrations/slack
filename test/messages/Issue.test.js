const { Issue } = require('../../lib/messages/issue');

const issuesOpened = require('../fixtures/webhooks/issues.opened.json');
const issuesClosed = require('../fixtures/webhooks/issues.closed.json');

describe('Issue rendering', () => {
  test('works for full unfurls', async () => {
    const issueMessage = new Issue({
      issue: issuesOpened.issue,
      repository: issuesOpened.repository,
      unfurlType: 'full',
    });
    const rendered = issueMessage.getRenderedMessage();
    expect(rendered).toMatchSnapshot();
  });

  test('works for condensed unfurls', async () => {
    const issueMessage = new Issue({
      issue: issuesOpened.issue,
      repository: issuesOpened.repository,
      unfurlType: 'condensed',
    });
    const rendered = issueMessage.getRenderedMessage();
    expect(rendered).toMatchSnapshot();
  });

  test('works for notification messages', async () => {
    const issueMessage = new Issue({
      issue: issuesOpened.issue,
      repository: issuesOpened.repository,
      eventType: 'issues.opened',
      sender: issuesOpened.sender,
    });
    const rendered = issueMessage.getRenderedMessage();
    expect(rendered).toMatchSnapshot();
  });

  test('works for minor notifications', async () => {
    const issueMessage = new Issue({
      issue: issuesClosed.issue,
      repository: issuesClosed.repository,
      eventType: 'issues.closed',
      sender: issuesClosed.sender,
    });
    const rendered = issueMessage.getRenderedMessage();
    expect(rendered).toMatchSnapshot();
  });

  test('extracts image from body_html', () => {
    const message = new Issue({
      issue: {
        ...issuesOpened.issue,
        body_html: '<p>Hello world!</p><p><img src="https://media.giphy.com/media/5xtDarEbygs3Pu7p3jO/giphy.gif"></p>',
      },
      repository: issuesOpened.repository,
      unfurlType: 'full',
    });
    expect(message.getRenderedMessage()).toMatchSnapshot();
  });

  test('does not render tables', async () => {
    const message = new Issue({
      issue: {
        ...issuesOpened.issue,
        body_html: '<p>Hi, this is my issue and below this line of text there is a table</p>\n<table>\n<thead>\n<tr>\n<th>test</th>\n<th>hello</th>\n<th>hi</th>\n<th>this is</th>\n<th>a table</th>\n</tr>\n</thead>\n<tbody>\n<tr>\n<td>with</td>\n<td>many</td>\n<td>test</td>\n<td>columns</td>\n<td>and</td>\n</tr>\n<tr>\n<td>rows</td>\n<td>ok</td>\n<td></td>\n<td></td>\n<td></td>\n</tr>\n<tr>\n<td></td>\n<td></td>\n<td></td>\n<td>third</td>\n<td>row!</td>\n</tr></tbody></table>',
      },
      repository: issuesOpened.repository,
      eventType: 'issues.opened',
    });
    expect(message.getRenderedMessage()).toMatchSnapshot();
  });
});
