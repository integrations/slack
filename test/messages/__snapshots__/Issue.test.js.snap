// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Issue rendering does not render tables 1`] = `
Object {
  "attachments": Array [
    Object {
      "author_icon": "https://avatars1.githubusercontent.com/u/7718702?v=4",
      "author_link": "https://github.com/wilhelmklopp",
      "author_name": "wilhelmklopp",
      "color": "#36a64f",
      "fallback": "[github-slack/public-test] Issue opened by wilhelmklopp",
      "fields": Array [
        Object {
          "short": true,
          "title": "Assignees",
          "value": "wilhelmklopp",
        },
        Object {
          "short": true,
          "title": "Labels",
          "value": "enhancement, question, wontfix",
        },
      ],
      "footer": "<https://github.com/github-slack/public-test|github-slack/public-test>",
      "footer_icon": "https://github.githubassets.com/favicon.ico",
      "mrkdwn_in": Array [
        "text",
      ],
      "pretext": "Issue opened by wilhelmklopp",
      "text": "Hi, this is my issue and below this line of text there is a table",
      "title": "#1 Needs content",
      "title_link": "https://github.com/github-slack/public-test/issues/1",
      "ts": 1508171038,
    },
  ],
}
`;

exports[`Issue rendering extracts image from body_html 1`] = `
Object {
  "attachments": Array [
    Object {
      "author_icon": "https://avatars1.githubusercontent.com/u/7718702?v=4",
      "author_link": "https://github.com/wilhelmklopp",
      "author_name": "wilhelmklopp",
      "color": "#36a64f",
      "fallback": "[github-slack/public-test] #1 Needs content",
      "fields": Array [
        Object {
          "short": true,
          "title": "Assignees",
          "value": "wilhelmklopp",
        },
        Object {
          "short": true,
          "title": "Labels",
          "value": "enhancement, question, wontfix",
        },
      ],
      "footer": "<https://github.com/github-slack/public-test|github-slack/public-test>",
      "footer_icon": "https://github.githubassets.com/favicon.ico",
      "image_url": "https://media.giphy.com/media/5xtDarEbygs3Pu7p3jO/giphy.gif",
      "mrkdwn_in": Array [
        "text",
      ],
      "text": "Hello world!

https://media.giphy.com/media/5xtDarEbygs3Pu7p3jO/giphy.gif",
      "title": "#1 Needs content",
      "title_link": "https://github.com/github-slack/public-test/issues/1",
      "ts": 1508171038,
    },
  ],
}
`;

exports[`Issue rendering works for condensed unfurls 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#36a64f",
      "fallback": "[github-slack/public-test] #1 Needs content",
      "footer": "<https://github.com/github-slack/public-test|github-slack/public-test>",
      "footer_icon": "https://github.githubassets.com/favicon.ico",
      "mrkdwn_in": Array [
        "text",
      ],
      "title": "#1 Needs content",
      "title_link": "https://github.com/github-slack/public-test/issues/1",
      "ts": 1508171038,
    },
  ],
}
`;

exports[`Issue rendering works for full unfurls 1`] = `
Object {
  "attachments": Array [
    Object {
      "author_icon": "https://avatars1.githubusercontent.com/u/7718702?v=4",
      "author_link": "https://github.com/wilhelmklopp",
      "author_name": "wilhelmklopp",
      "color": "#36a64f",
      "fallback": "[github-slack/public-test] #1 Needs content",
      "fields": Array [
        Object {
          "short": true,
          "title": "Assignees",
          "value": "wilhelmklopp",
        },
        Object {
          "short": true,
          "title": "Labels",
          "value": "enhancement, question, wontfix",
        },
      ],
      "footer": "<https://github.com/github-slack/public-test|github-slack/public-test>",
      "footer_icon": "https://github.githubassets.com/favicon.ico",
      "mrkdwn_in": Array [
        "text",
      ],
      "text": "This repo needs some content, as currently, it does not have **any**.",
      "title": "#1 Needs content",
      "title_link": "https://github.com/github-slack/public-test/issues/1",
      "ts": 1508171038,
    },
  ],
}
`;

exports[`Issue rendering works for minor notifications 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#cb2431",
      "fallback": "[github-slack/public-test] Issue closed by bkeepers",
      "footer": "<https://github.com/github-slack/public-test|github-slack/public-test>",
      "footer_icon": "https://github.githubassets.com/favicon.ico",
      "mrkdwn_in": Array [
        "text",
      ],
      "pretext": "Issue closed by bkeepers",
      "title": "#1 Needs content",
      "title_link": "https://github.com/github-slack/public-test/issues/1",
      "ts": 1508171038,
    },
  ],
}
`;

exports[`Issue rendering works for notification messages 1`] = `
Object {
  "attachments": Array [
    Object {
      "author_icon": "https://avatars1.githubusercontent.com/u/7718702?v=4",
      "author_link": "https://github.com/wilhelmklopp",
      "author_name": "wilhelmklopp",
      "color": "#36a64f",
      "fallback": "[github-slack/public-test] Issue opened by wilhelmklopp",
      "fields": Array [
        Object {
          "short": true,
          "title": "Assignees",
          "value": "wilhelmklopp",
        },
        Object {
          "short": true,
          "title": "Labels",
          "value": "enhancement, question, wontfix",
        },
      ],
      "footer": "<https://github.com/github-slack/public-test|github-slack/public-test>",
      "footer_icon": "https://github.githubassets.com/favicon.ico",
      "mrkdwn_in": Array [
        "text",
      ],
      "pretext": "Issue opened by wilhelmklopp",
      "text": "This repo needs some content, as currently, it does not have **any**.",
      "title": "#1 Needs content",
      "title_link": "https://github.com/github-slack/public-test/issues/1",
      "ts": 1508171038,
    },
  ],
}
`;
